File: v2/src/main/io/export.ts
import { EJSON } from 'bson';
import { createWriteStream } from 'node:fs';
import type { MongoClient } from 'mongodb';
import type { ExportRequest, IoResult } from '../../shared/io.js';
import { parseFilter, parseProjection, parseSort } from '../mongo/queryParser.js';

export async function exportFind(
  client: MongoClient,
  req: ExportRequest,
): Promise<IoResult> {
  const t0 = performance.now();
  try {
    const filter = parseFilter(req.filter);
    const projection = parseProjection(req.projection);
    const sort = parseSort(req.sort);

    let cursor = client.db(req.db).collection(req.collection).find(filter);
    if (Object.keys(projection).length > 0) cursor = cursor.project(projection);
    if (Object.keys(sort).length > 0) cursor = cursor.sort(sort);
    if ((req.limit ?? 0) > 0) cursor = cursor.limit(req.limit!);

    const out = createWriteStream(req.filePath);
    let written = 0;

    if (req.format === 'json') {
      out.write('[\n');
    }

    let csvHeaders: string[] | null = null;

    for await (const doc of cursor) {
      const ejson = EJSON.serialize(doc as object, { relaxed: false });
      if (req.format === 'ndjson') {
        out.write(JSON.stringify(ejson));
        out.write('\n');
      } else if (req.format === 'json') {
        if (written > 0) out.write(',\n');
        out.write(JSON.stringify(ejson));
      } else if (req.format === 'csv') {
        if (csvHeaders === null) {
          csvHeaders = Object.keys(ejson as Record<string, unknown>);
          out.write(csvHeaders.map(csvQuote).join(','));
          out.write('\n');
        }
        const row = csvHeaders
          .map((h) => csvQuote(stringifyCell((ejson as Record<string, unknown>)[h])))
          .join(',');
        out.write(row);
        out.write('\n');
      }
      written++;
    }

    if (req.format === 'json') {
      out.write('\n]\n');
    }

    await new Promise<void>((resolve, reject) => {
      out.end((err: unknown) => (err ? reject(err as Error) : resolve()));
    });

    return {
      ok: true,
      written,
      durationMs: Math.round(performance.now() - t0),
    };
  } catch (err) {
    return {
      ok: false,
      error: err instanceof Error ? err.message : String(err),
      durationMs: Math.round(performance.now() - t0),
    };
  }
}

function csvQuote(s: string): string {
  if (/[",\n\r]/.test(s)) {
    return `"${s.replace(/"/g, '""')}"`;
  }
  return s;
}

function stringifyCell(v: unknown): string {
  if (v === undefined || v === null) return '';
  if (typeof v === 'string') return v;
  return JSON.stringify(v);
}

File: v3/src/main/kotlin/io/mex/io/Export.kt
package io.mex.io

import com.mongodb.client.MongoClient
import io.mex.mongo.parseFilter
import io.mex.mongo.parseProjection
import io.mex.mongo.parseSort
import org.bson.Document
import org.bson.json.JsonMode
import org.bson.json.JsonWriterSettings
import java.io.BufferedWriter
import java.nio.file.Files
import java.nio.file.Path

enum class ExportFormat { json, ndjson, csv }

data class ExportRequest(
    val connectionId: String,
    val db: String,
    val collection: String,
    val format: ExportFormat,
    val filter: String = "",
    val projection: String = "",
    val sort: String = "",
    val limit: Int = 0,
    val target: Path,
)

data class ExportResult(
    val ok: Boolean,
    val written: Long = 0,
    val durationMs: Long,
    val error: String? = null,
)

private val EJSON: JsonWriterSettings = JsonWriterSettings.builder().outputMode(JsonMode.EXTENDED).build()

fun exportFind(client: MongoClient, req: ExportRequest): ExportResult {
    val t0 = System.nanoTime()
    return runCatching {
        val filter = parseFilter(req.filter)
        val projection = parseProjection(req.projection)
        val sort = parseSort(req.sort)
        var cursor = client.getDatabase(req.db).getCollection(req.collection).find(filter)
        if (!projection.isEmpty) cursor = cursor.projection(projection)
        if (!sort.isEmpty) cursor = cursor.sort(sort)
        if (req.limit > 0) cursor = cursor.limit(req.limit)

        Files.newBufferedWriter(req.target).use { out ->
            var written = 0L
            var headers: List<String>? = null
            if (req.format == ExportFormat.json) out.write("[\n")

            for (doc in cursor) {
                when (req.format) {
                    ExportFormat.ndjson -> { out.write(doc.toJson(EJSON)); out.write("\n") }
                    ExportFormat.json -> {
                        if (written > 0) out.write(",\n")
                        out.write(doc.toJson(EJSON))
                    }
                    ExportFormat.csv -> {
                        if (headers == null) {
                            headers = doc.keys.toList()
                            out.write(headers!!.joinToString(",") { csvQuote(it) })
                            out.write("\n")
                        }
                        out.write(headers!!.joinToString(",") { key -> csvQuote(stringifyCell(doc[key])) })
                        out.write("\n")
                    }
                }
                written++
            }
            if (req.format == ExportFormat.json) out.write("\n]\n")
            ExportResult(ok = true, written = written, durationMs = (System.nanoTime() - t0) / 1_000_000)
        }
    }.getOrElse {
        ExportResult(ok = false, error = it.message, durationMs = (System.nanoTime() - t0) / 1_000_000)
    }
}

private fun csvQuote(s: String): String {
    return if (Regex("[\",\\n\\r]").containsMatchIn(s)) "\"${s.replace("\"", "\"\"")}\"" else s
}

private fun stringifyCell(v: Any?): String = when (v) {
    null -> ""
    is String -> v
    is Document -> v.toJson(EJSON)
    is List<*> -> v.toString()
    else -> v.toString()
}

private fun BufferedWriter.write(d: Document) = this.write(d.toJson(EJSON))
